### PR TITLE
Fix the effective constraints so that they don't exceed max values

### LINKED ISSUE #84224: effectiveConstraints is buggy.
I added a visual density to my app, and now I get:

BoxConstraints has both width and height constraints non-normalized

The problem is I'm setting a BoxContraint to 32,32 (min/max for width and height) on an IconButton

But internally it calls effectiveContraints and changes the minWidth, but it doesn't set the maxWidth.

After this call, I have a boxContraint that is minWidth: 44 and maxWidth: 32 and this triggers the crash.

I'm on Flutter web if that matters, but I assume it would happen everywhere.
```dart
/// Return a copy of [constraints] whose minimum width and height have been
  /// updated with the [baseSizeAdjustment].
  BoxConstraints effectiveConstraints(BoxConstraints constraints){
    assert(constraints != null && constraints.debugAssertIsValid());
    return constraints.copyWith(
      minWidth: (constraints.minWidth + baseSizeAdjustment.dx).clamp(0.0, double.infinity).toDouble(),
      minHeight: (constraints.minHeight + baseSizeAdjustment.dy).clamp(0.0, double.infinity).toDouble(),
    );
  }
```
I'm doing this:
```dart
  IconButton(
      constraints: BoxConstraints.tight(const Size(32, 32)),
    ...
```
And I added this for the visualDensity to my app
```dart 
      ThemeData(visualDensity: Utils.isWeb
          ? const VisualDensity(horizontal: 3, vertical: 3)
          : VisualDensity.adaptivePlatformDensity,
```

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This fixes `VisualDensity.effectiveConstraints` to clamp to the max values of the constraint, instead of to `double.infinity`.

## Related Issues

- Fixes https://github.com/flutter/flutter/issues/84224

## Tests

- Added tests for `effectiveConstraints`.

### CODE REVIEW

File: packages/flutter/lib/src/material/theme_data.dart [hunk 1]
```diff
@@ -2270,11 +2270,14 @@ class VisualDensity with Diagnosticable {
 
   /// Return a copy of [constraints] whose minimum width and height have been
   /// updated with the [baseSizeAdjustment].
+  ///
+  /// The resulting minWidth and minHeight values are clamped to not exceed the
+  /// maxWidth and maxHeight values, respectively.
   BoxConstraints effectiveConstraints(BoxConstraints constraints){
     assert(constraints != null && constraints.debugAssertIsValid());
     return constraints.copyWith(
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/theme_data.dart:2278]: It looks likes this should be:
```dart
return constraints.constrain(constraints.smallest + baseSizeAdjustment);
```

What's the difference?

Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart:2278]: The difference is that constrain returns a `Size`, and we need to return updated constraints.

Comment (HansMuller) [reply packages/flutter/lib/src/material/theme_data.dart:2278]: Sorry, I wasn't paying attention. The idea was to write an expression where it was obvious that we were applying the existing constraints to the adjusted version.  There doesn't seem to be an obvious way to do that with the BoxConstraints methods.

File: packages/flutter/test/material/theme_data_test.dart [hunk 2]
```diff
@@ -203,6 +203,49 @@ void main() {
     }
   }, variant: TargetPlatformVariant.all());
 
+  testWidgets('Ensure Visual Density effective constraints are clamped', (WidgetTester tester) async {
+    const BoxConstraints square = BoxConstraints.tightFor(width: 35.0, height: 35.0);
```
Comment (HansMuller) [inline packages/flutter/test/material/theme_data_test.dart:207]: Best to use 35 instead of 35.0, just for the sake of consistency.

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**